src/PMD/parser/parser.jl

```diff
@@ -0,0 +1,523 @@
+
+        return parse!(parser)
+    end
+
+    return data_struct
+end
+
+function parse!(parser::Parser)
+    for line in readlines(parser.io)
+        parser.line += 1
+
+        # TODO: Remove comments properly
+        line = strip(line)
+
+        if isempty(line) || startswith(line, "//")
+            continue # comment or empty line
+        end
+
+        _parse_line!(parser, line, _get_state(parser))
+    end
+
+    if !(_get_state(parser) isa PMD_IDLE)
+        _syntax_error(parser, "Unexpected EOF")
+    end
+
+    # apply merges
+    for collection in keys(parser.data_struct)
+        _apply_merge!(parser, collection, String[collection])
+    end
+
+    # delete temporary classes (starting with "_")
+    for collection in keys(parser.data_struct)
+        if startswith(collection, "_MERGE")
+            delete!(parser.data_struct, collection)
+        end
+    end
+
+    return nothing
+end
+
+function _parse_line!(parser::Parser, line::AbstractString, ::PMD_IDLE)
+    # Look for model definition block
+    m = match(r"DEFINE_MODEL\s+(MODL:)?(\S+)", line)
+
+    if !isnothing(m)
+        model_name = String(m[2])
+
+        if _hasinv(parser.model_template, model_name)
+            collection = parser.model_template.inv[model_name]
+        else
+            _syntax_warning(parser, "Unknown model '$(model_name)'")
+
+            _push_state!(parser, PMD_DEF_MODEL(nothing))
+
+            return nothing
+        end
+
+        parser.data_struct[collection] = Dict{String, Attribute}()
+
+        # default attributes that belong to "all collections"
+        parser.data_struct[collection]["name"] = Attribute("name", false, String, 0, "")
+        parser.data_struct[collection]["code"] = Attribute("code", false, Int32, 0, "")
+        parser.data_struct[collection]["AVId"] = Attribute("AVId", false, String, 0, "")
+
+        if collection == "PSRSystem"
+            parser.data_struct[collection]["id"] = Attribute("id", false, String, 0, "")
+        end
+
+        _push_state!(parser, PMD_DEF_MODEL(collection))
+
+        return nothing
+    end
+
+    # Look for class definition block
+    m = match(r"DEFINE_CLASS\s+(\S+)", line)
+
+    if !isnothing(m)
+        collection = String(m[1])
+
+        parser.data_struct[collection] = Dict{String, Attribute}()
+
+        # default attributes that belong to "all collections"
+        parser.data_struct[collection]["name"] = Attribute("name", false, String, 0, "")
+        parser.data_struct[collection]["code"] = Attribute("code", false, Int32, 0, "")
+        parser.data_struct[collection]["AVId"] = Attribute("AVId", false, String, 0, "")
+
+        if collection == "PSRSystem"
+            parser.data_struct[collection]["id"] = Attribute("id", false, String, 0, "")
+        end
+
+        _push_state!(parser, PMD_DEF_CLASS(collection))
+
+        return nothing
+    end
+
+    # Look for class merge block
+    m = match(r"MERGE_CLASS\s+(\S+)\s+(\S+)", line)
+
+    if !isnothing(m)
```

Comment:
We could be consistent in the nothing checks, below you made `state.collection === nothing`

As a personal preference I prefer `x === nothing` and `x !== nothing`

Reply:
This is prior to a discussion about this with @joaquimg.
We're sticking to `===`

Reply:
This is done in the parser context, but there are other `isnothing` occurences in other places.